src/server/journal/serializer.cc

```diff
@@ -0,0 +1,204 @@
+  auto [cmd, tail_args] = args;
+
+  RETURN_ON_ERR(WriteU64(1 + tail_args.size()));
+  RETURN_ON_ERR(Write(cmd));
+  for (auto v : tail_args)
+    RETURN_ON_ERR(Write(v));
+
+  return std::error_code{};
+}
+error_code JournalWriter::Write(CmdArgList args) {
+  RETURN_ON_ERR(WriteU64(args.size()));
+  for (auto v : args)
+    RETURN_ON_ERR(Write(facade::ToSV(v)));
+
+  return std::error_code{};
+}
+
+error_code JournalWriter::Write(std::monostate) {
+  return std::error_code{};
+}
+
+error_code JournalWriter::Write(const journal::EntryNew& entry) {
+  if (entry.opcode != journal::Op::SELECT && (!cur_dbid_ || entry.dbid != *cur_dbid_)) {
+    RETURN_ON_ERR(Write(journal::EntryNew{journal::Op::SELECT, entry.dbid}));
+    cur_dbid_ = entry.dbid;
+  }
+
+  RETURN_ON_ERR(WriteU8(uint8_t(entry.opcode)));
+
+  switch (entry.opcode) {
+    case journal::Op::SELECT:
+      return WriteU16(entry.dbid);
+    case journal::Op::VAL:
+      RETURN_ON_ERR(WriteU64(entry.txid));
+      return std::visit([this](const auto& payload) { return Write(payload); }, entry.payload);
+    default:
+      break;
+  };
+  return std::error_code{};
+}
+
+JournalReader::JournalReader(io::Source* source, DbIndex dbid)
+    : source_{source}, buf_{}, dbid_{dbid} {
+}
+
+io::Result<uint8_t> JournalReader::ReadU8() {
+  uint8_t buf[1];
+  size_t read;
+  SET_OR_UNEXPECT(source_->Read(io::MutableBytes{buf}), read);
+  if (read < 1)  // TODO: Custom errc namespace? Generic opcodes?
+    return make_unexpected(std::make_error_code(std::errc::result_out_of_range));
+  return buf[0];
+}
+
+io::Result<uint16_t> JournalReader::ReadU16() {
+  uint8_t buf[2];
+  size_t read;
+  SET_OR_UNEXPECT(source_->Read(io::MutableBytes{buf}), read);
+  if (read < 2)  // TODO: Custom errc namespace? Generic opcodes?
+    return make_unexpected(std::make_error_code(std::errc::result_out_of_range));
+  return absl::big_endian::Load16(buf);
+}
+
+io::Result<uint64_t> JournalReader::ReadU64() {
+  uint8_t buf[8];
+  size_t read;
+  SET_OR_UNEXPECT(source_->Read(io::MutableBytes{buf}), read);
+  if (read < 8)  // TODO: Custom errc namespace? Generic opcodes?
+    return make_unexpected(std::make_error_code(std::errc::result_out_of_range));
+  return absl::big_endian::Load64(buf);
+}
+
+io::Result<size_t> JournalReader::ReadString() {
+  size_t size = 0;
+  SET_OR_UNEXPECT(ReadU64(), size);
+
+  buf_.Reserve(buf_.InputLen() + size);  // TODO: There should be a more intuitive function
+
+  auto dest = buf_.AppendBuffer().first(size);
+  auto res = source_->Read(dest);
+  if (!res)
+    return res;
+
+  buf_.CommitWrite(*res);
+  if (*res != size)
+    return make_unexpected(std::make_error_code(std::errc::message_size));
+
+  return size;
+}
+
+std::error_code JournalReader::Read(CmdArgVec* vec) {
+  buf_.ConsumeInput(buf_.InputBuffer().size());
+
+  size_t size = 0;
+  SET_OR_RETURN(ReadU64(), size);
+
+  vec->resize(size);
+  for (auto& span : *vec) {
+    size_t len;
+    SET_OR_RETURN(ReadString(), len);
```

Comment:
you can change ReadString to return IoBuf::Bytes and this way you dont need the second loop

Reply:
I need it, because I can't keep any references to it while its growing

Reply:
I see, please consider writing the cmd length in the writer so you could use reserve on the buffer at the beginning of the loop